File: src/main/resources/assets/sectionController.js
var enabled = false;
var Sections = {
    UNDEFINED: -1,

    TOP: 1,
    TOP_LEFT: 2,
    TOP_RIGHT: 3,

    RIGHT: 4,
    LEFT: 5,

    BOTTOM: 6,
    BOTTOM_LEFT: 7,
    BOTTOM_RIGHT: 8,

    CENTER: 9,
};

var controlEdgeWidth = 100;
var controlEdgeStepSize = 10;

var currentSection = Sections.UNDEFINED;
var previousSection = Sections.UNDEFINED;

var intervalID;

var resetSections = function(event) {
    if (enabled) {
        currentSection = Sections.UNDEFINED;
        previousSection = Sections.UNDEFINED;

        clearInterval(intervalID);
    }

}
var sectionController = function(event) {
    if (enabled) {
        var x = event.pageX,
            y = event.pageY;
        var mousePosition = new ClientCoordinate(x, y);

        setCurrentSection(mousePosition);

        if (currentSection == Sections.CENTER) {
            clearInterval(intervalID);
        }

        if (previousSection != currentSection) {
            previousSection = currentSection;

            if (currentSection != Sections.CENTER) {
                clearInterval(intervalID);
                intervalID = setInterval(function() {
                    var offsetY = controlEdgeStepSize * scale;
                    var offsetX = controlEdgeStepSize * scale;

                    if (currentSection == Sections.TOP) {
                        corner.y -= offsetY;
                    } else if (currentSection == Sections.TOP_LEFT) {
                        corner.x -= offsetX;
                        corner.y -= offsetY;
                    } else if (currentSection == Sections.TOP_RIGHT) {
                        corner.x += offsetX;
                        corner.y -= offsetY;
                    } else if (currentSection == Sections.LEFT) {
                        corner.x -= offsetX;
                    } else if (currentSection == Sections.RIGHT) {
                        corner.x += offsetX;
                    } else if (currentSection == Sections.BOTTOM) {
                        corner.y += offsetY;
                    } else if (currentSection == Sections.BOTTOM_LEFT) {
                        corner.x -= offsetX;
                        corner.y += offsetY;
                    } else if (currentSection == Sections.BOTTOM_RIGHT) {
                        corner.x += offsetX;
                        corner.y += offsetY;
                    }

                    drawScene();
                }, 20);
            }
        }
    }
}

function setCurrentSection(mousePosition) {
    if (onTop(mousePosition)) {
        if (onLeft(mousePosition)) {
            currentSection = Sections.TOP_LEFT;
        } else if (onRight(mousePosition)) {
            currentSection = Sections.TOP_RIGHT;
        } else {
            currentSection = Sections.TOP;
        }
    } else if (onBottom(mousePosition)) {
        if (onLeft(mousePosition)) {
            currentSection = Sections.BOTTOM_LEFT;
        } else if (onRight(mousePosition)) {
            currentSection = Sections.BOTTOM_RIGHT;
        } else {
            currentSection = Sections.BOTTOM;
        }
    } else if (onLeft(mousePosition)) {
        currentSection = Sections.LEFT;
    } else if (onRight(mousePosition)) {
        currentSection = Sections.RIGHT;
    } else {
        currentSection = Sections.CENTER;
    }
}

function onLeft(coordinate) {
    return coordinate.x <= controlEdgeWidth;
}

function onRight(coordinate) {
    return coordinate.x >= width - controlEdgeWidth;
}

function onTop(coordinate) {
    return coordinate.y <= controlEdgeWidth;
}

function onBottom(coordinate) {
    return coordinate.y >= height - controlEdgeWidth;
}

File: frontend/src/main/kotlin/com/github/pozo/SectionController.kt
package com.github.pozo

import org.w3c.dom.events.MouseEvent
import kotlin.browser.window

enum class Sections {
    UNDEFINED,
    TOP,
    TOP_LEFT,
    TOP_RIGHT,

    RIGHT,
    LEFT,

    BOTTOM,
    BOTTOM_LEFT,
    BOTTOM_RIGHT,

    CENTER
}

class SectionController(val screen: Screen, val canvas: Canvas) {
    var enabled = false
    var controlEdgeWidth = 100
    var controlEdgeStepSize = 10

    var currentSection = Sections.UNDEFINED
    var previousSection = Sections.UNDEFINED

    var intervalID: Int = 0
    var interval = 30

    fun resetSections() {
        if (enabled) {
            currentSection = Sections.UNDEFINED
            previousSection = Sections.UNDEFINED

            window.clearInterval(intervalID)
        }
    }

    fun control(event: MouseEvent) {
        if (enabled) {
            val x = event.pageX
            val y = event.pageY
            val mousePosition = Coordinate(x.toInt(), y.toInt())

            setCurrentSection(mousePosition)

            if (currentSection == Sections.CENTER) {
                window.clearInterval(intervalID)
            }

            if (previousSection != currentSection) {
                previousSection = currentSection

                if (currentSection != Sections.CENTER) {
                    window.clearInterval(intervalID)
                    intervalID = window.setInterval({
                        val screenCorner = screen.screenCorner
                        val currentX = screenCorner.x
                        val currentY = screenCorner.y

                        val offsetY = controlEdgeStepSize * screen.scale
                        val offsetX = controlEdgeStepSize * screen.scale

                        if (currentSection == Sections.TOP) {
                            screen.setScreenCornerY((currentY - offsetY).toInt())
                        } else if (currentSection == Sections.TOP_LEFT) {
                            screen.setScreenCornerX((currentX - offsetX).toInt())
                            screen.setScreenCornerY((currentY - offsetY).toInt())
                        } else if (currentSection == Sections.TOP_RIGHT) {
                            screen.setScreenCornerX((currentX + offsetX).toInt())
                            screen.setScreenCornerY((currentY - offsetY).toInt())
                        } else if (currentSection == Sections.LEFT) {
                            screen.setScreenCornerX((currentX - offsetX).toInt())
                        } else if (currentSection == Sections.RIGHT) {
                            screen.setScreenCornerX((currentX + offsetX).toInt())
                        } else if (currentSection == Sections.BOTTOM) {
                            screen.setScreenCornerY((currentY + offsetY).toInt())
                        } else if (currentSection == Sections.BOTTOM_LEFT) {
                            screen.setScreenCornerX((currentX - offsetX).toInt())
                            screen.setScreenCornerY((currentY + offsetY).toInt())
                        } else if (currentSection == Sections.BOTTOM_RIGHT) {
                            screen.setScreenCornerX((currentX + offsetX).toInt())
                            screen.setScreenCornerY((currentY + offsetY).toInt())
                        }
                        canvas.drawScene()
                    }, interval)
                }
            }
        }
    }

    fun setCurrentSection(mousePosition: Coordinate) {
        if (onTop(mousePosition)) {
            if (onLeft(mousePosition)) {
                currentSection = Sections.TOP_LEFT
            } else if (onRight(mousePosition)) {
                currentSection = Sections.TOP_RIGHT
            } else {
                currentSection = Sections.TOP
            }
        } else if (onBottom(mousePosition)) {
            if (onLeft(mousePosition)) {
                currentSection = Sections.BOTTOM_LEFT
            } else if (onRight(mousePosition)) {
                currentSection = Sections.BOTTOM_RIGHT
            } else {
                currentSection = Sections.BOTTOM
            }
        } else if (onLeft(mousePosition)) {
            currentSection = Sections.LEFT
        } else if (onRight(mousePosition)) {
            currentSection = Sections.RIGHT
        } else {
            currentSection = Sections.CENTER
        }
    }

    fun onLeft(coordinate: Coordinate): Boolean {
        return coordinate.x <= controlEdgeWidth
    }

    fun onRight(coordinate: Coordinate): Boolean {
        return coordinate.x >= screen.screenWidth - controlEdgeWidth
    }

    fun onTop(coordinate: Coordinate): Boolean {
        return coordinate.y <= controlEdgeWidth
    }

    fun onBottom(coordinate: Coordinate): Boolean {
        return coordinate.y >= screen.screenHeight - controlEdgeWidth
    }
}


